gtsfm/averaging/rotation/uncertainity_aware_rotation_averaging.py

```diff
@@ -0,0 +1,257 @@
+"""
+Uncertainity aware rotation averaging.
+
+This algorithm was proposed in "Revisiting Rotation Averaging: Uncertainties and Robust Losses" and is implemented by 
+following the authors' source code.
+
+References:
+- https://openaccess.thecvf.com/content/CVPR2023/papers/Zhang_Revisiting_Rotation_Averaging_Uncertainties_and_Robust_Losses_CVPR_2023_paper.pdf
+- https://github.com/zhangganlin/GlobalSfMpy
+"""
+import math
+from enum import Enum
+from typing import Dict, List, NamedTuple, Optional, Tuple
+
+import numpy as np
+import pyceres
+from gtsam import Rot3
+import networkx as nx
+from gtsfm.common.pose_prior import PosePrior
+
+import gtsfm.utils.logger as logger_utils
+import gtsfm.averaging.rotation.const as rotation_const
+from gtsfm.averaging.rotation.rotation_averaging_base import RotationAveragingBase
+
+logger = logger_utils.get_logger()
+
+DUMMY_COVARIANCE = np.eye(3) * 1e-2
+
+
+def random_rotation() -> Rot3:
+    """Sample a random rotation by generating a sample from the 4d unit sphere."""
+    q = np.random.randn(4) * 0.03
+    # make unit-length quaternion
+    q /= np.linalg.norm(q)
+    qw, qx, qy, qz = q
+    R = Rot3(qw, qx, qy, qz)
+    return R
+
+
+def initialize_global_rotations_using_mst(num_images: int, i2Ri1_dict: Dict[Tuple[int, int], Rot3]) -> List[Rot3]:
+    # Create a graph from the relative rotations dictionary
+    graph = nx.Graph()
+    for i1, i2 in i2Ri1_dict.keys():
+        # TODO: use inlier count as weight
+        graph.add_edge(i1, i2, weight=1)
+
+    # Compute the Minimum Spanning Tree (MST)
+    mst = nx.minimum_spanning_tree(graph)
+
+    wRis = [random_rotation() for _ in range(num_images)]
+    for i1, i2 in sorted(mst.edges):
+        if (i1, i2) in i2Ri1_dict:
+            wRis[i2] = wRis[i1] * i2Ri1_dict[(i1, i2)].inverse()
+        else:
+            wRis[i2] = wRis[i1] * i2Ri1_dict[(i2, i1)]
+
+    return wRis
+
+
+class UncertainityAwareRotationAveraging(RotationAveragingBase):
+    def run_rotation_averaging(
+        self,
+        num_images: int,
+        i2Ri1_dict: Dict[Tuple[int, int], Optional[Rot3]],
+        i1Ti2_priors: Dict[Tuple[int, int], PosePrior],
+    ) -> List[Optional[Rot3]]:
+        rotation_info_dict: Dict[Tuple[int, int], RotationInfo] = {
+            k: RotationInfo(i2Ri1=v, covariance_mat=DUMMY_COVARIANCE) for k, v in i2Ri1_dict.items() if v is not None
+        }
+
+        global_rotations_init = initialize_global_rotations_using_mst(
+            num_images, {key: value for key, value in i2Ri1_dict.items() if value is not None}
+        )
+        # global_rotations_init = [random_rotation() for _ in range(num_images)]
+
+        # return global_rotations_init
+
+        return _estimate_rotations_with_customized_loss_and_covariance_ceres(global_rotations_init, rotation_info_dict)
+
+
+class RotationInfo(NamedTuple):
```

Comment:
nit: can we add an `Attributes` section to describe each field here, and a 1-line sentence describing the class?